tools/ete/ete_lineage_generator.py

```diff
@@ -0,0 +1,129 @@
+import optparse
+import sys
+
+from ete3 import NCBITaxa
+
+# - compared to gi2taxonomy the root is excluded, since
+#   the value is always "root", i.e. useless information
+# - additional levels that appear in the ncbi taxdb have
+#   been added
+#   (order from https://en.wikipedia.org/wiki/Taxonomic_rank#All_ranks)
+# TODO the full list of ranks could be derived from the input DB
+LONG_RANKS = [u"superkingdom", u"kingdom", u"subkingdom",
+              u"superphylum", u"phylum", u"subphylum",
+              u"superclass", u"class", u"subclass", "infraclass",
+              u"cohort",
+              u"superorder", u"order", u"suborder", u"infraorder", u"parvorder",
+              u"superfamily", u"family", u"subfamily",
+              u"tribe", u"subtribe",
+              u"genus", u"subgenus",
+              u"species group", u"species subgroup", u"species", u"subspecies",
+              u"varietas", "forma"]
+
+SHORT_RANKS = [u"kingdom",
+               u"phylum",
+               u"class",
+               u"order",
+               u"family",
+               u"genus",
+               u"species"]
+
+
+def process_taxid(ncbi, taxid, ranks, RANK_IDX, lower=False):
+    """
+    process one taxid:
+        - get lineage (as list of taxids, ranks, and names)
+        - reverse the lineage if lower ranks are to be used for filling
+        - fill the ranks with the data from the lineage
+    ncbi: ete NCBITaxa object
+    taxid: a taxid (int)
+    ranks: list of ranks (should be initialized with "NA" x number of levels of interest)
+    RANK_IDX: mapping from rank names to indices (distance to root/leaf?)
+    lower: use lower taxa for filling "NA"s
+    """
+    lineage = ncbi.get_lineage(taxid)
+    lineage_ranks = ncbi.get_rank(lineage)
+    lineage_names = ncbi.get_taxid_translator(lineage, try_synonyms=True)
+    if lower:
+        lineage.reverse()
+    for l in lineage:
+        if not lineage_ranks[l] in RANK_IDX:
+            continue
+        if ranks[ RANK_IDX[lineage_ranks[l]]] != "NA":
```

Comment:
Can you remove the space between `[` and `RANK_IDX` to make flake8 happy, please?